Refs #25520 -- Tests for IntegrityErrors deleting M2M to proxy models

jmurty opened the pull request:
Add unit tests that demonstrate the error condition where deleting a
proxy-based object fails if that object has reverse M2M relationships
to a proxy ancestor, using a database that enforces foreign key
integrity constraints (e.g. PostgreSQL, not SQLite)

The root cause is Django does not find/include reverse M2M relationships
to proxy models when collecting objects for deletion via
`get_deleted_objects` etc, these relationships are only included if they
target concrete parent models.

In the new tests:
- two of the four tests pass, because the M2M relationship targets a
  concrete model
- the other two tests fail with integrity errors, because the M2M rel
  targets a proxy model.

jmurty commented:
This is a re-submission of PR #5399 rebased onto master instead of the 1.7.x branch

timgraham commented:
The issue seems valid, however, linking it to the polymorphic ticket doesn't seem appropriate. I'd open a separate ticket for clarity. Probably the tests that already pass aren't needed as I suspect these cases are already tested.

jmurty commented:
I have created a new ticket that is more specific to this issue and changed the Ref in this PR's title: https://code.djangoproject.com/ticket/25520

And removed the passing tests.

jmurty commented:
Added a potential fix. It relies on a new Options method `get_proxy_parent_list` that will need some sanity-checking, but which might be useful for related tickets and in general.

timgraham commented:
buildbot, add to whitelist.

charettes commented:
I don't need we need such a complex test case requiring the creation of more than 5 extra tables.

Since m2m fields are really just foreign keys under the hood [the simple test](https://code.djangoproject.com/attachment/ticket/25505/25505-reproduced-issue.diff) provided for [#25505](https://code.djangoproject.com/ticket/25505) should be enough.

timgraham commented:
Closing in favor of https://github.com/django/django/pull/5378.
